Hashie::Extensions::DeepLocate

msievers opened the pull request:
Hi everybody, I would like to propose the following extension

```ruby
Hashie::Extensions::DeepLocate
```

It's similar to Hashie::Extensions::DeepFind, but differs in two mayor aspects
* it uses a *comparator callable* instead of a key
* it returns the container enumerables for found elements

I needed this to be able to do modify some complex elasticsearch search requests. These are plain hashes, but the queries can be nested in many ways. One does not know where exactly a certain element might be located. So I wanted to be able to search using some advanced comparator and to modify the results later on in place, which can only be accomplished if the container objects are returned.

**Example**
```ruby
books = [
  {
    title: "Ruby for beginners",
    pages: 120
  },
  {
    title: "CSS for intermediates",
    pages: 80
  },
  {
    title: "Collection of ruby books",
    books: [
      {
        title: "Ruby for the rest of us",
        pages: 576
      }
    ]
  }
]

books.extend(Hashie::Extensions::DeepLocate)

books.deep_locate -> (key, value, object) { key == :title && value.include?("Ruby") }
# => [{:title=>"Ruby for beginners", :pages=>120}, {:title=>"Ruby for the rest of us", :pages=>576}]

books.deep_locate -> (key, value, object) { key == :pages && value <= 120 }
# => [{:title=>"Ruby for beginners", :pages=>120}, {:title=>"CSS for intermediates", :pages=>80}]
```
Please see the specs for some more advanced examples.

gregory commented:
i like it! 
What's the `object`? books? or the current "sub-hash"? 
Also, can we include this module? (haven't looked at the code)

msievers commented:
```object``` is the current Enumerable. This means that for the books example, it would be the books array for the first level. This way one can make decision based on the whole Enumerable, rather then only on the current (key)/value.

gregory commented:
yeah that's what i thought. also looks like it's includable :) 
Feature looks good to me, can't check the code right now. maybe really quick tonight. 
thx!

msievers commented:
I try to fix the travis (ruby 1.9/lambda) issues asap.

msievers commented:
Travis is green now. It seems that there are no spaces allowed in lambda hash rocket syntax prior to ruby 2.0. I removed the spaces and left a note in the docs.

dblock commented:
Just for the sake of the argument, why isn't `DeepFind` extended to support this? Have you tried?

msievers commented:
The semantics of the present DeepFind are different and I didn't want to break anything.

DeepFind returns the matching member values of an Enumerable, whereas DeepLocate returns the Enumerable itself, making it possible to modify the found Enumerable in place, which isn't possible with DeepFind under any circumstances.

Taking the doc example of DeepFind

```ruby
options = {
  user: {
    location: {
      address: '123 Street',
      zip_code: 12345
    }
  }
}
options.deep_find(:address) # => '123 Street' # the member value
options.deep_locate(:address) # => { address: '123 Street', zip_code: 12345 } # the complete location hash

# deep_locate(:symbol) is a shortcut which creates a key comparator on-the-fly
```

dblock commented:
Can we implement `DeepFind` on top of `DeepLocate`?

msievers commented:
This should be fairly easy. Its just accessing the given key on the result. For example it may look something like this.

```ruby
options = {
  user: {
    location: {
      address: '123 Street',
      zip_code: 12345
    }
  }
}

def options.deep_find_all(key)
  deep_locate(key).map do |result|
    result[key]
  end
end

def options.deep_find(key)
  deep_find_all(key).first
end

options.deep_find(:address) # => '123 Street'
```

dblock commented:
I think you should do it part of this PR, if all specs pass I say this will be a go.

msievers commented:
Alright, I will do this tomorrow.

msievers commented:
Hi! Compared to yesterday I made the following changes

* moved main functionality into a module method ```Hashie::Extensions::DeepLocate.deep_locate```
  * the instance method now only calls it with ```object = self```
  * this makes it possible to even use DeepLocate without including/extend the original hash
* changed specs accordingly
* added specs for including/extending classes/objects
* implemented DeepFind on top of DeepLocate
  * I intentionally followed the original structure as far as possible, so it should not break even if somebody relies on the original private methods

dblock commented:
This is perfect, merging. Nice work @msievers.
